CreateMultiStargateAndDatacenter intermittently fails

### What happened?

The test `CreateMultiStargateAndDatacenter` [fails intermittently](https://github.com/k8ssandra/k8ssandra-operator/runs/4999948818?check_suite_focus=true#logs). 

(Full logs reproduced below as they'll be deleted eventually from GHA.)

### Logs:

```
/usr/local/bin/kubectl --context kind-k8ssandra-0 -n multi-stargate-lm72lm apply -f /home/runner/work/k8ssandra-operator/k8ssandra-operator/test/testdata/fixtures/multi-stargate
Error:     suite_test.go:761: check that datacenter dc1 in cluster kind-k8ssandra-0 is ready
Error:     suite_test.go:792: check that Stargate s1 in cluster kind-k8ssandra-0 is ready
Error:     suite_test.go:535: retrieve database credentials
Error:     suite_test.go:539: deploying Stargate ingress routes in kind-k8ssandra-0
/usr/local/bin/kustomize edit set namespace multi-stargate-lm72lm
time="2022-01-31T01:11:27Z" level=info msg="kustomize build | kubectl apply" Context=kind-k8ssandra-0 Dir=../../build/test-config/ingress/kind-k8ssandra-0 Namespace=multi-stargate-lm72lm
/usr/local/bin/kustomize build
/usr/local/bin/kubectl --context kind-k8ssandra-0 -n multi-stargate-lm72lm apply --server-side --force-conflicts -f -
=== RUN   TestOperator/CreateMultiStargateAndDatacenter/TestStargateApis[0]
=== RUN   TestOperator/CreateMultiStargateAndDatacenter/TestStargateApis[0]/TestStargateNativeApi
Error:     stargate_apis_test.go:25: test Stargate native API in context kind-k8ssandra-0
    stargate_apis_test.go:257: 
        	Error Trace:	stargate_apis_test.go:257
        	            				stargate_apis_test.go:51
        	            				stargate_apis_test.go:26
        	Error:      	Object expected to be of type *message.SchemaChangeResult, but was *message.ServerError
        	Test:       	TestOperator/CreateMultiStargateAndDatacenter/TestStargateApis[0]/TestStargateNativeApi
        	Messages:   	Expected CREATE TABLE response to be of type SchemaChangeResult
1:12AM ERR CQL client conn [L:127.0.0.1:43468 <-> R:127.0.0.1:30942]: server replied with fatal error code ErrorCode ServerError [0x00000000], closing connection
=== RUN   TestOperator/CreateMultiStargateAndDatacenter/TestStargateApis[0]/TestStargateRestApi
Error:     stargate_apis_test.go:29: test Stargate REST API in context kind-k8ssandra-0
=== RUN   TestOperator/CreateMultiStargateAndDatacenter/TestStargateApis[0]/TestStargateRestApi/TestSchemaApi
=== RUN   TestOperator/CreateMultiStargateAndDatacenter/TestStargateApis[0]/TestStargateRestApi/TestDocumentApi
/usr/local/bin/kubectl --context kind-k8ssandra-0 -n multi-stargate-lm72lm delete IngressRoute --all
/usr/local/bin/kubectl --context kind-k8ssandra-0 -n multi-stargate-lm72lm delete IngressRouteTCP --all
time="2022-01-31T01:12:50Z" level=info msg="dumping cluster info"
time="2022-01-31T01:12:54Z" level=info msg="Deleting CassandraDatacenters" Namespace=multi-stargate-lm72lm
time="2022-01-31T01:13:07Z" level=info msg="deleting namespace" Context=kind-k8ssandra-0 Namespace=multi-stargate-lm72lm
time="2022-01-31T01:13:07Z" level=info msg="deleting namespace" Context=kind-k8ssandra-1 Namespace=multi-stargate-lm72lm
time="2022-01-31T01:13:17Z" level=info msg="waiting for namespace deletion" error=null
time="2022-01-31T01:13:27Z" level=info msg="waiting for namespace deletion" error=null
time="2022-01-31T01:13:37Z" level=info msg="waiting for namespace deletion" error=null
time="2022-01-31T01:13:47Z" level=info msg="waiting for namespace deletion" error=null
time="2022-01-31T01:13:57Z" level=info msg="deleting namespace" Context=kind-k8ssandra-1 Namespace=multi-stargate-lm72lm
time="2022-01-31T01:13:57Z" level=info msg="deleting namespace" Context=kind-k8ssandra-0 Namespace=multi-stargate-lm72lm
--- FAIL: TestOperator (798.33s)
    --- FAIL: TestOperator/CreateMultiStargateAndDatacenter (798.33s)
        --- FAIL: TestOperator/CreateMultiStargateAndDatacenter/TestStargateApis[0] (73.90s)
            --- FAIL: TestOperator/CreateMultiStargateAndDatacenter/TestStargateApis[0]/TestStargateNativeApi (33.68s)
            --- PASS: TestOperator/CreateMultiStargateAndDatacenter/TestStargateApis[0]/TestStargateRestApi (40.22s)
                --- PASS: TestOperator/CreateMultiStargateAndDatacenter/TestStargateApis[0]/TestStargateRestApi/TestSchemaApi (9.24s)
                --- PASS: TestOperator/CreateMultiStargateAndDatacenter/TestStargateApis[0]/TestStargateRestApi/TestDocumentApi (30.49s)
FAIL
FAIL	github.com/k8ssandra/k8ssandra-operator/test/e2e	798.362s
FAIL
Error: make: *** [Makefile:126: e2e-test] Error 1
Error: Process completed with exit code 2.
```

--- a/CHANGELOG/CHANGELOG-1.0.md
+++ b/CHANGELOG/CHANGELOG-1.0.md
@@ -18,6 +18,7 @@ When cutting a new release, update the `unreleased` heading to the tag being gen
 * [FEATURE] [#296](https://github.com/k8ssandra/k8ssandra-operator/issues/296) Expose the stopped property from the
   CassandraDatacenter spec
 * [TESTING] [#332](https://github.com/k8ssandra/k8ssandra-operator/issues/332) Configure mocks to avoid panics
+* [TESTING] [#361](https://github.com/k8ssandra/k8ssandra-operator/issues/361) Reduce flakiness in Stargate API tests
 
 ## v0.4.0 2022-02-04
 

--- a/test/e2e/stargate_apis_test.go
+++ b/test/e2e/stargate_apis_test.go
@@ -20,7 +20,7 @@ import (
 )
 
 func testStargateApis(t *testing.T, ctx context.Context, k8sContextName string, k8sContextIdx int, username string, password string, replication map[string]int) {
-	t.Run(fmt.Sprintf("TestStargateApis[%d]", k8sContextIdx), func(t *testing.T) {
+	t.Run(fmt.Sprintf("TestStargateApis[%s]", k8sContextName), func(t *testing.T) {
 		t.Run("TestStargateNativeApi", func(t *testing.T) {
 			t.Log("test Stargate native API in context " + k8sContextName)
 			testStargateNativeApi(t, ctx, k8sContextIdx, username, password, replication)
@@ -70,49 +70,54 @@ func testDocumentApi(t *testing.T, restClient *resty.Client, k8sContextIdx int, 
 }
 
 func createKeyspaceAndTableRest(t *testing.T, restClient *resty.Client, k8sContextIdx int, token, tableName, keyspaceName string, replication map[string]int) {
-	keyspaceUrl := fmt.Sprintf("http://stargate.127.0.0.1.nip.io:3%v080/v2/schemas/keyspaces", k8sContextIdx)
-	keyspaceJson := fmt.Sprintf(`{"name":"%v","datacenters":%v}`, keyspaceName, formatReplicationForRestApi(replication))
-	request := restClient.NewRequest().
-		SetHeader("Content-Type", "application/json").
-		SetHeader("X-Cassandra-Token", token).
-		SetBody(keyspaceJson)
-	response, err := request.Post(keyspaceUrl)
-	require.NoError(t, err, "Create keyspace with Schema API failed")
-	assert.Equal(t, http.StatusCreated, response.StatusCode(), "Expected create keyspace request to return 201")
 	timeout := 2 * time.Minute
 	interval := 1 * time.Second
 	require.Eventually(t, func() bool {
-		keyspaceUrl = fmt.Sprintf("http://stargate.127.0.0.1.nip.io:3%v080/v2/schemas/keyspaces/%v", k8sContextIdx, keyspaceName)
-		request = restClient.NewRequest().
+		keyspaceUrl := fmt.Sprintf("http://stargate.127.0.0.1.nip.io:3%v080/v2/schemas/keyspaces", k8sContextIdx)
+		keyspaceJson := fmt.Sprintf(`{"name":"%v","datacenters":%v}`, keyspaceName, formatReplicationForRestApi(replication))
+		request := restClient.NewRequest().
 			SetHeader("Content-Type", "application/json").
-			SetHeader("X-Cassandra-Token", token)
-		response, err = request.Get(keyspaceUrl)
-		return err == nil && response.StatusCode() == http.StatusOK
-	}, timeout, interval)
-	tableUrl := fmt.Sprintf("http://stargate.127.0.0.1.nip.io:3%v080/v2/schemas/keyspaces/%s/tables", k8sContextIdx, keyspaceName)
-	tableJson := fmt.Sprintf(`{ "name": "%v",
-  "columnDefinitions": [
-    { "name": "pk", "typeDefinition": "int", "static": false },
-    { "name": "cc", "typeDefinition": "int", "static": false },
-    { "name": "v" , "typeDefinition": "int", "static": false }
-  ],
-  "primaryKey": { "partitionKey": [ "pk" ], "clusteringKey": [ "cc" ] }
-}`, tableName)
-	request = restClient.NewRequest().
-		SetHeader("Content-Type", "application/json").
-		SetHeader("X-Cassandra-Token", token).
-		SetBody(tableJson)
-	response, err = request.Post(tableUrl)
-	require.NoError(t, err, "Create table with Schema API failed")
-	assert.Equal(t, http.StatusCreated, response.StatusCode(), "Expected create table request to return 201")
+			SetHeader("X-Cassandra-Token", token).
+			SetBody(keyspaceJson)
+		response, err := request.Post(keyspaceUrl)
+		return err == nil && response.StatusCode() == http.StatusCreated
+	}, timeout, interval, "Create keyspace with Schema API failed")
 	require.Eventually(t, func() bool {
-		tableUrl = fmt.Sprintf("http://stargate.127.0.0.1.nip.io:3%v080/v2/schemas/keyspaces/%v/tables/%v", k8sContextIdx, keyspaceName, tableName)
-		request = restClient.NewRequest().
+		keyspaceUrl := fmt.Sprintf("http://stargate.127.0.0.1.nip.io:3%v080/v2/schemas/keyspaces/%v", k8sContextIdx, keyspaceName)
+		request := restClient.NewRequest().
 			SetHeader("Content-Type", "application/json").
 			SetHeader("X-Cassandra-Token", token)
-		response, err = request.Get(tableUrl)
+		response, err := request.Get(keyspaceUrl)
 		return err == nil && response.StatusCode() == http.StatusOK
-	}, timeout, interval)
+	}, timeout, interval, "Retrieve keyspace with Schema API failed")
+	require.Eventually(t, func() bool {
+		tableUrl := fmt.Sprintf("http://stargate.127.0.0.1.nip.io:3%v080/v2/schemas/keyspaces/%s/tables", k8sContextIdx, keyspaceName)
+		tableJson := fmt.Sprintf(`{ "name": "%v",
+		  "columnDefinitions": [
+			{ "name": "pk", "typeDefinition": "int", "static": false },
+			{ "name": "cc", "typeDefinition": "int", "static": false },
+			{ "name": "v" , "typeDefinition": "int", "static": false }
+		  ],
+		  "primaryKey": { "partitionKey": [ "pk" ], "clusteringKey": [ "cc" ] }
+		}`, tableName)
+		request := restClient.NewRequest().
+			SetHeader("Content-Type", "application/json").
+			SetHeader("X-Cassandra-Token", token).
+			SetBody(tableJson)
+		_, err := request.Post(tableUrl)
+		// We don't test the status code as it could be 500 if the table already exists because it was created in a
+		// previous call, but somehow the HTTP request failed, and now we are attempting to create it again; we will
+		// validate below that the table exists.
+		return err == nil
+	}, timeout, interval, "Create table with Schema API failed")
+	require.Eventually(t, func() bool {
+		tableUrl := fmt.Sprintf("http://stargate.127.0.0.1.nip.io:3%v080/v2/schemas/keyspaces/%v/tables/%v", k8sContextIdx, keyspaceName, tableName)
+		request := restClient.NewRequest().
+			SetHeader("Content-Type", "application/json").
+			SetHeader("X-Cassandra-Token", token)
+		response, err := request.Get(tableUrl)
+		return err == nil && response.StatusCode() == http.StatusOK
+	}, timeout, interval, "Retrieve table with Schema API failed")
 }
 
 func insertRowsRest(t *testing.T, restClient *resty.Client, k8sContextIdx int, token string, nbRows int, tableName, keyspaceName string) {
@@ -142,33 +147,35 @@ func checkRowCountRest(t *testing.T, restClient *resty.Client, k8sContextIdx int
 		SetHeader("Content-Type", "application/json").
 		SetHeader("X-Cassandra-Token", token)
 	response, err := request.Get(tableUrl + params.Encode())
-	assert.NoError(t, err, "Retrieve rows with Schema API failed")
-	assert.Equal(t, http.StatusOK, response.StatusCode(), "Expected retrieve data request to return 200")
-	data := string(response.Body())
-	expected := fmt.Sprintf(`"count":%v`, nbRows)
-	assert.Contains(t, data, expected, "Expected response body to contain count:%d", nbRows)
+	if assert.NoError(t, err, "Retrieve rows with Schema API failed") &&
+		assert.Equal(t, http.StatusOK, response.StatusCode(), "Expected retrieve data request to return 200") {
+		data := string(response.Body())
+		expected := fmt.Sprintf(`"count":%v`, nbRows)
+		assert.Contains(t, data, expected, "Expected response body to contain count:%d", nbRows)
+	}
 }
 
 func createDocumentNamespace(t *testing.T, restClient *resty.Client, k8sContextIdx int, token, documentNamespace string, replication map[string]int) string {
-	url := fmt.Sprintf("http://stargate.127.0.0.1.nip.io:3%v080/v2/schemas/namespaces", k8sContextIdx)
-	documentNamespaceJson := fmt.Sprintf(`{"name":"%s","datacenters":%v}`, documentNamespace, formatReplicationForRestApi(replication))
-	request := restClient.NewRequest().
-		SetHeader("Content-Type", "application/json").
-		SetHeader("X-Cassandra-Token", token).
-		SetBody(documentNamespaceJson)
-	response, err := request.Post(url)
-	require.NoError(t, err, "Failed creating Stargate document namespace")
-	assert.Equal(t, http.StatusCreated, response.StatusCode(), "Expected create namespace request to return 201")
 	timeout := 2 * time.Minute
 	interval := 1 * time.Second
 	require.Eventually(t, func() bool {
-		url = fmt.Sprintf("http://stargate.127.0.0.1.nip.io:3%v080/v2/schemas/namespaces/%v", k8sContextIdx, documentNamespace)
-		request = restClient.NewRequest().
+		url := fmt.Sprintf("http://stargate.127.0.0.1.nip.io:3%v080/v2/schemas/namespaces", k8sContextIdx)
+		documentNamespaceJson := fmt.Sprintf(`{"name":"%s","datacenters":%v}`, documentNamespace, formatReplicationForRestApi(replication))
+		request := restClient.NewRequest().
+			SetHeader("Content-Type", "application/json").
+			SetHeader("X-Cassandra-Token", token).
+			SetBody(documentNamespaceJson)
+		response, err := request.Post(url)
+		return err == nil && response.StatusCode() == http.StatusCreated
+	}, timeout, interval, "Create namespace with Document API failed")
+	require.Eventually(t, func() bool {
+		url := fmt.Sprintf("http://stargate.127.0.0.1.nip.io:3%v080/v2/schemas/namespaces/%v", k8sContextIdx, documentNamespace)
+		request := restClient.NewRequest().
 			SetHeader("Content-Type", "application/json").
 			SetHeader("X-Cassandra-Token", token)
-		response, err = request.Get(url)
+		response, err := request.Get(url)
 		return err == nil && response.StatusCode() == http.StatusOK
-	}, timeout, interval)
+	}, timeout, interval, "Retrieve namespace with Document API failed")
 	return documentNamespace
 }
 
@@ -177,7 +184,7 @@ const (
 	awesomeMovieName     = "Watchmen"
 )
 
-func writeDocument(t *testing.T, restClient *resty.Client, k8sContextIdx int, token, documentNamespace, documentId string) string {
+func writeDocument(t *testing.T, restClient *resty.Client, k8sContextIdx int, token, documentNamespace, documentId string) {
 	url := fmt.Sprintf("http://stargate.127.0.0.1.nip.io:3%v080/v2/namespaces/%s/collections/movies/%s", k8sContextIdx, documentNamespace, documentId)
 	awesomeMovieDocument := map[string]string{"Director": awesomeMovieDirector, "Name": awesomeMovieName}
 	response, err := restClient.NewRequest().
@@ -185,11 +192,11 @@ func writeDocument(t *testing.T, restClient *resty.Client, k8sContextIdx int, to
 		SetHeader("X-Cassandra-Token", token).
 		SetBody(awesomeMovieDocument).
 		Put(url)
-	assert.NoError(t, err, "Failed writing Stargate document")
-	stargateResponse := string(response.Body())
-	expectedResponse := fmt.Sprintf("{\"documentId\":\"%s\"}", documentId)
-	assert.Equal(t, expectedResponse, stargateResponse, "Unexpected response from Stargate: '%s'", stargateResponse)
-	return documentId
+	if assert.NoError(t, err, "Failed writing Stargate document") {
+		stargateResponse := string(response.Body())
+		expectedResponse := fmt.Sprintf("{\"documentId\":\"%s\"}", documentId)
+		assert.Equal(t, expectedResponse, stargateResponse, "Unexpected response from Stargate: '%s'", stargateResponse)
+	}
 }
 
 func readDocument(t *testing.T, restClient *resty.Client, k8sContextIdx int, token, documentNamespace, documentId string) {
@@ -201,30 +208,38 @@ func readDocument(t *testing.T, restClient *resty.Client, k8sContextIdx int, tok
 	require.NoError(t, err, "Failed to retrieve Stargate document")
 	var genericJson map[string]interface{}
 	err = json.Unmarshal(response.Body(), &genericJson)
-	require.NoError(t, err, "Failed to decode Stargate response")
-	assert.Equal(t, documentId, genericJson["documentId"], "Expected JSON payload to contain a field documentId")
-	require.NotNil(t, genericJson["data"], "Expected JSON payload to contain a field data")
-	require.IsType(t, map[string]interface{}{}, genericJson["data"], "Expected field data to be map[string]interface{}")
-	assert.Equal(t, awesomeMovieDirector, genericJson["data"].(map[string]interface{})["Director"], "Expected JSON payload to contain a field data.Director")
-	assert.Equal(t, awesomeMovieName, genericJson["data"].(map[string]interface{})["Name"], "Expected JSON payload to contain a field data.Name")
+	if assert.NoError(t, err, "Failed to decode Stargate response") &&
+		assert.Contains(t, genericJson, "documentId") &&
+		assert.Contains(t, genericJson, "data") {
+		assert.Equal(t, documentId, genericJson["documentId"], "Expected JSON payload to contain a field documentId")
+		if assert.IsType(t, map[string]interface{}{}, genericJson["data"], "Expected field data to be map[string]interface{}") {
+			data := genericJson["data"].(map[string]interface{})
+			assert.Equal(t, awesomeMovieDirector, data["Director"], "Expected JSON payload to contain a field data.Director")
+			assert.Equal(t, awesomeMovieName, data["Name"], "Expected JSON payload to contain a field data.Name")
+		}
+	}
 }
 
 func authenticate(t *testing.T, restClient *resty.Client, k8sContextIdx int, username, password string) string {
-	url := fmt.Sprintf("http://stargate.127.0.0.1.nip.io:3%v080/v1/auth", k8sContextIdx)
-	body := map[string]string{"username": username, "password": password}
-	request := restClient.NewRequest().
-		SetHeader("Content-Type", "application/json").
-		SetBody(body)
-	response, err := request.Post(url)
-	require.NoError(t, err, "Authentication via REST API failed")
-	require.Equal(t, http.StatusCreated, response.StatusCode(), "Expected auth request to return 201")
 	var result map[string]interface{}
-	err = json.Unmarshal(response.Body(), &result)
-	require.NoError(t, err, "Failed to decode Stargate auth response")
+	require.Eventually(t, func() bool {
+		url := fmt.Sprintf("http://stargate.127.0.0.1.nip.io:3%v080/v1/auth", k8sContextIdx)
+		body := map[string]string{"username": username, "password": password}
+		request := restClient.NewRequest().
+			SetHeader("Content-Type", "application/json").
+			SetBody(body)
+		response, err := request.Post(url)
+		return err == nil &&
+			response.StatusCode() == http.StatusCreated &&
+			json.Unmarshal(response.Body(), &result) == nil
+	}, time.Minute, time.Second, "Authentication via REST failed")
 	token, found := result["authToken"]
 	assert.True(t, found, "REST authentication response did not have expected authToken field")
-	tokenStr := token.(string)
-	assert.NotEmpty(t, tokenStr, "REST authentication response did not have expected authToken field")
+	var tokenStr string
+	if assert.IsType(t, "", token) {
+		tokenStr = token.(string)
+		assert.NotEmpty(t, tokenStr, "REST authentication response did not have expected authToken field")
+	}
 	return tokenStr
 }
 
@@ -236,48 +251,59 @@ func openCqlClientConnection(t *testing.T, ctx context.Context, k8sContextIdx in
 	}
 	cqlClient := client.NewCqlClient(contactPoint, credentials)
 	cqlClient.ConnectTimeout = 30 * time.Second
-	cqlClient.ReadTimeout = 3 * time.Minute
+	cqlClient.ReadTimeout = 1 * time.Minute
 	connection, err := cqlClient.ConnectAndInit(ctx, primitive.ProtocolVersion4, client.ManagedStreamId)
 	require.NoError(t, err, "Failed to connect via CQL native port to %s", contactPoint)
 	return connection
 }
 
 func createKeyspaceAndTableNative(t *testing.T, connection *client.CqlClientConnection, tableName, keyspaceName string, replication map[string]int) {
-	response := sendQuery(t, connection, fmt.Sprintf(
-		"CREATE KEYSPACE IF NOT EXISTS %s with replication = {'class':'NetworkTopologyStrategy', %s}",
-		keyspaceName,
-		formatReplicationForCql(replication),
-	))
-	require.IsType(t, &message.SchemaChangeResult{}, response.Body.Message, "Expected CREATE KEYSPACE response to be of type SchemaChangeResult")
-	response = sendQuery(t, connection, fmt.Sprintf(
-		"CREATE TABLE IF NOT EXISTS %s.%s (id timeuuid PRIMARY KEY, val text)",
-		keyspaceName,
-		tableName,
-	))
-	require.IsType(t, &message.SchemaChangeResult{}, response.Body.Message, "Expected CREATE TABLE response to be of type SchemaChangeResult")
+	require.Eventually(t, func() bool {
+		response, err := sendQuery(connection, fmt.Sprintf(
+			"CREATE KEYSPACE IF NOT EXISTS %s with replication = {'class':'NetworkTopologyStrategy', %s}",
+			keyspaceName,
+			formatReplicationForCql(replication),
+		))
+		if err != nil {
+			return false
+		}
+		_, ok := response.Body.Message.(*message.SchemaChangeResult)
+		return ok
+	}, time.Minute, time.Second, "CREATE KEYSPACE via CQL failed")
+	require.Eventually(t, func() bool {
+		response, err := sendQuery(connection, fmt.Sprintf(
+			"CREATE TABLE IF NOT EXISTS %s.%s (id timeuuid PRIMARY KEY, val text)",
+			keyspaceName,
+			tableName,
+		))
+		if err != nil {
+			return false
+		}
+		_, ok := response.Body.Message.(*message.SchemaChangeResult)
+		return ok
+	}, time.Minute, time.Second, "CREATE TABLE via CQL failed")
 }
 
 func insertRowsNative(t *testing.T, connection *client.CqlClientConnection, nbRows int, tableName, keyspaceName string) {
 	for i := 0; i < nbRows; i++ {
-		response := sendQuery(t, connection, fmt.Sprintf(
-			"INSERT INTO %s.%s (id, val) VALUES (now(), '%d')",
-			keyspaceName,
-			tableName,
-			i,
-		))
+		query := fmt.Sprintf("INSERT INTO %s.%s (id, val) VALUES (now(), '%d')", keyspaceName, tableName, i)
+		response, err := sendQuery(connection, query)
+		assert.NoError(t, err, "Query failed: %s", query)
 		assert.IsType(t, &message.VoidResult{}, response.Body.Message, "Expected INSERT INTO response to be of type VoidResult")
 	}
 }
 
 func checkRowCountNative(t *testing.T, connection *client.CqlClientConnection, nbRows int, tableName, keyspaceName string) {
-	response := sendQuery(t, connection, fmt.Sprintf("SELECT id FROM %s.%s", keyspaceName, tableName))
-	if assert.IsType(t, &message.RowsResult{}, response.Body.Message, "Expected SELECT response to be of type RowsResult") {
+	query := fmt.Sprintf("SELECT id FROM %s.%s", keyspaceName, tableName)
+	response, err := sendQuery(connection, query)
+	if assert.NoError(t, err, "Query failed: %s", query) &&
+		assert.IsType(t, &message.RowsResult{}, response.Body.Message, "Expected SELECT response to be of type RowsResult") {
 		result := response.Body.Message.(*message.RowsResult)
 		assert.Len(t, result.Data, nbRows, "Expected SELECT query to return %d rows", nbRows)
 	}
 }
 
-func sendQuery(t *testing.T, connection *client.CqlClientConnection, query string) *frame.Frame {
+func sendQuery(connection *client.CqlClientConnection, query string) (*frame.Frame, error) {
 	request := frame.NewFrame(
 		primitive.ProtocolVersion4,
 		client.ManagedStreamId,
@@ -288,10 +314,7 @@ func sendQuery(t *testing.T, connection *client.CqlClientConnection, query strin
 			},
 		},
 	)
-	response, err := connection.SendAndReceive(request)
-	require.NoError(t, err, "Failed to send QUERY request: %s", query)
-	require.NotNil(t, response.Body, "Expected non nil response body")
-	return response
+	return connection.SendAndReceive(request)
 }
 
 func formatReplicationForRestApi(replication map[string]int) string {

--- a/test/e2e/stargate_apis_test.go
+++ b/test/e2e/stargate_apis_test.go
@@ -92,7 +92,10 @@ func createKeyspaceAndTableRest(t *testing.T, restClient *resty.Client, k8sConte
 	}, timeout, interval, "Retrieve keyspace with Schema API failed")
 	require.Eventually(t, func() bool {
 		tableUrl := fmt.Sprintf("http://stargate.127.0.0.1.nip.io:3%v080/v2/schemas/keyspaces/%s/tables", k8sContextIdx, keyspaceName)
-		tableJson := fmt.Sprintf(`{ "name": "%v",
+		tableJson := fmt.Sprintf(
+			`{ 
+          "name": "%v", 
+          "ifNotExists": true, 
 		  "columnDefinitions": [
 			{ "name": "pk", "typeDefinition": "int", "static": false },
 			{ "name": "cc", "typeDefinition": "int", "static": false },
@@ -105,8 +108,7 @@ func createKeyspaceAndTableRest(t *testing.T, restClient *resty.Client, k8sConte
 			SetHeader("X-Cassandra-Token", token).
 			SetBody(tableJson)
 		response, err := request.Post(tableUrl)
-		t.Logf("err = %v code = %v", err, response.StatusCode())
-		return err == nil && response.StatusCode() == http.StatusOK
+		return err == nil && response.StatusCode() == http.StatusCreated
 	}, timeout, interval, "Create table with Schema API failed")
 	require.Eventually(t, func() bool {
 		tableUrl := fmt.Sprintf("http://stargate.127.0.0.1.nip.io:3%v080/v2/schemas/keyspaces/%v/tables/%v", k8sContextIdx, keyspaceName, tableName)

--- a/test/e2e/stargate_apis_test.go
+++ b/test/e2e/stargate_apis_test.go
@@ -104,11 +104,8 @@ func createKeyspaceAndTableRest(t *testing.T, restClient *resty.Client, k8sConte
 			SetHeader("Content-Type", "application/json").
 			SetHeader("X-Cassandra-Token", token).
 			SetBody(tableJson)
-		_, err := request.Post(tableUrl)
-		// We don't test the status code as it could be 500 if the table already exists because it was created in a
-		// previous call, but somehow the HTTP request failed, and now we are attempting to create it again; we will
-		// validate below that the table exists.
-		return err == nil
+		response, err := request.Post(tableUrl)
+		return err == nil && response.StatusCode() == http.StatusOK
 	}, timeout, interval, "Create table with Schema API failed")
 	require.Eventually(t, func() bool {
 		tableUrl := fmt.Sprintf("http://stargate.127.0.0.1.nip.io:3%v080/v2/schemas/keyspaces/%v/tables/%v", k8sContextIdx, keyspaceName, tableName)

--- a/test/e2e/stargate_apis_test.go
+++ b/test/e2e/stargate_apis_test.go
@@ -105,6 +105,7 @@ func createKeyspaceAndTableRest(t *testing.T, restClient *resty.Client, k8sConte
 			SetHeader("X-Cassandra-Token", token).
 			SetBody(tableJson)
 		response, err := request.Post(tableUrl)
+		t.Logf("err = %v code = %v", err, response.StatusCode())
 		return err == nil && response.StatusCode() == http.StatusOK
 	}, timeout, interval, "Create table with Schema API failed")
 	require.Eventually(t, func() bool {

